usr/share/rear/prep/BAREOS/default/560_check_bareos_filesets.sh

```diff
@@ -1,26 +1,91 @@
+#      purgedfiles: 0
+#             type: B
+#            level: F
+#         clientid: 2
+#           client: client1
+#        jobstatus: T
+#        schedtime: 2024-05-29 16:03:46
+#        starttime: 2024-05-29 16:03:49
+#          endtime: 2024-05-29 16:05:34
+#      realendtime: 2024-05-29 16:05:34
+#         duration: 00:01:45
+#         jobtdate: 1,716,998,734
+#     volsessionid: 3
+#   volsessiontime: 1,716,996,683
+#         jobfiles: 43,785
+#         jobbytes: 2,007,335,334
+#        joberrors: 0
+#  jobmissingfiles: 0
+#           poolid: 3
+#         poolname: Full
+#       priorjobid: 0
+#        filesetid: 2
+#          fileset: LinuxAll
+# 
+#            jobid: 4
+#              job: backup-client1.2024-05-29_16.41.30_21
+#             name: backup-client1
+#      purgedfiles: 0
+#             type: B
+#            level: I
+#         clientid: 2
+#           client: client1
+#        jobstatus: T
+#        schedtime: 2024-05-29 16:41:28
+#        starttime: 2024-05-29 16:41:32
+#          endtime: 2024-05-29 16:41:35
+#      realendtime: 2024-05-29 16:41:35
+#         duration: 00:00:03
+#         jobtdate: 1,717,000,895
+#     volsessionid: 4
+#   volsessiontime: 1,716,996,683
+#         jobfiles: 2,792
+#         jobbytes: 310,942,237
+#        joberrors: 0
+#  jobmissingfiles: 0
+#           poolid: 2
+#         poolname: Incremental
+#       priorjobid: 0
+#        filesetid: 2
+#          fileset: LinuxAll
 
-# if we have more then 1 fileset for a client then we need also to define variable BAREOS_FILESET
+mapfile -t filesets < <( bcommand "llist backups client=$BAREOS_CLIENT" | bcommand_extract_value "fileset" | sort | uniq )
 
-# Save the found fileset names in a file $TMP_DIR/bareos_filesets
-echo "show filesets" | bconsole | grep "Name =" | grep $HOSTNAME | cut -d= -f2 > "$TMP_DIR/bareos_filesets"
+LogPrint "available filesets:" "${filesets[@]}"
 
-# when amount of lines > 1 in file $TMP_DIR/bareos_filesets then we may decide that there is more then 1 fileset
-# for current host
+if (( ${#filesets[@]} == 0 )); then
+    Error "No valid backups found for client $BAREOS_CLIENT"
+fi
 
-nr_of_filesets=( $(wc -l $TMP_DIR/bareos_filesets) )
+if [ "$BAREOS_FILESET" ]; then
+    LogPrint "BAREOS_FILESET=$BAREOS_FILESET"
+    if ! IsInArray "$BAREOS_FILESET" "${filesets[@]}"; then
+        Error "No valid backup for fileset ($BAREOS_FILESET). Successful backups for following filesets: " "{filesets[@]}"
+    fi
+    return
+fi
 
-case "$nr_of_filesets" in
-    0 ) Error "No fileset defined in Bareos for $HOSTNAME" ;;
-    1 ) BAREOS_FILESET=""
-        Log "We found Bareos fileset : $(sed -e 's/"//g' $TMP_DIR/bareos_filesets)" ;;
-    * ) LogPrint "We found several defined Bareos filesets for $HOSTNAME :"
-        LogPrint "$( cat $TMP_DIR/bareos_filesets | sed -e 's/"//g' )"
-        Error "Define variable BAREOS_FILESET in $CONFIG_DIR/local.conf" ;;
-esac
+if (( ${#filesets[@]} == 1 )); then
```

Comment:
If this is the typical thing to happen then I'd assume that we should inform users somewhere to keep running `rear mkrescue` often enough to catch changing restore job details.

Reply:
I added a hint to the log message.